// @sudoku/stores/game.js
import { decodeSencode, encodeSudoku } from '@sudoku/sencode';
import { generateSudoku } from '@sudoku/sudoku';
import { createGame, createSudoku } from '../../../domain/index.js';
import { derived, get, writable } from 'svelte/store';

export const gamePaused = writable(true);

function createDomainGameStore() {
	const state = writable(createStateFromGrid(generateSudoku('easy')));

	function createStateFromGrid(puzzleGrid) {
		const sudoku = createSudoku(puzzleGrid);
		const game = createGame({ sudoku });
		return snapshotState(game);
	}

	function snapshotState(game) {
		const sudoku = game.getSudoku();
		return {
			game,
			sudoku, // Added this to expose domain object methods to UI
			puzzleGrid: sudoku.getPuzzleGrid(),
			userGrid: sudoku.getGrid(),
			invalidCells: sudoku.getInvalidCells(),
			canUndo: game.canUndo(),
			canRedo: game.canRedo(),
			isSolved: sudoku.isSolved(),
		};
	}

	function commitGame(game) {
		state.set(snapshotState(game));
	}

	return {
		subscribe: state.subscribe,

		startNew(diff = 'easy') {
			state.set(createStateFromGrid(generateSudoku(diff)));
		},

		startCustom(sencode) {
			state.set(createStateFromGrid(decodeSencode(sencode)));
		},

		guess(move) {
			state.update(($state) => {
				const changed = $state.game.guess(move);
				return changed ? snapshotState($state.game) : $state;
			});
		},

		undo() {
			state.update(($state) => {
				const changed = $state.game.undo();
				return changed ? snapshotState($state.game) : $state;
			});
		},

		redo() {
			state.update(($state) => {
				const changed = $state.game.redo();
				return changed ? snapshotState($state.game) : $state;
			});
		},

		getSencode() {
			const current = get(state);
			return encodeSudoku(current.puzzleGrid);
		},

		replacePuzzleGrid(puzzleGrid) {
			const sudoku = createSudoku(puzzleGrid);
			const game = createGame({ sudoku });
			commitGame(game);
		},
	};
}

export const domainGame = createDomainGameStore();
export const puzzleGrid = derived(domainGame, ($game) => $game.puzzleGrid);
export const userGrid = derived(domainGame, ($game) => $game.userGrid);
export const invalidCells = derived(domainGame, ($game) => $game.invalidCells);
export const canUndo = derived(domainGame, ($game) => $game.canUndo);
export const canRedo = derived(domainGame, ($game) => $game.canRedo);
export const gameWon = derived(domainGame, ($game) => $game.isSolved, false);
